### PR TITLE
Fix launching studio from quizzes

### PR DESCRIPTION
**Pull request by vanwagonet**
refs: MBL-14002
affects: Student
release note: Fixed an issue launching Studio videos from quizzes.

### CODE REVIEW

**Comment by instructure-cx** [comment]
Scan submitted to Checkmarx

File: Core/Core/Router/Router.swift [hunk 1]
```diff
@@ -152,7 +152,7 @@ public class Router: RouterProtocol {
     private func cleanURL(_ url: URLComponents) -> URLComponents {
         // URLComponents does all the encoding we care about except we often have + meaning space in query
         var url = url
-        url.query = url.query?.replacingOccurrences(of: "+", with: " ")
+        url.percentEncodedQuery = url.percentEncodedQuery?.replacingOccurrences(of: "+", with: "%20")
```
Comment (vanwagonet) [inline Core/Core/Router/Router.swift]: This was causing some decoding of `%26` to `&` within a query param, which then caused it to move up to be a new query param. Using the `percentEncodedQuery` fixes this.

File: Core/CoreTests/Extensions/URLComponentsExtensionsTests.swift [hunk 2]
```diff
@@ -31,6 +31,12 @@ class URLComponentsExtensionsTests: XCTestCase {
         XCTAssertEqual(URLComponents.parse(url), URLComponents(string: string)!)
     }
 
+    func testParseURLWithRecursion() {
+        let string = "scheme://h/p?u=https%3A%2F%2Fh%3Fu%3Dhttps%253A%252F%252Fh%253Fa%253Db%2526b%253Dc"
+        let url = URL(string: string)!
+        XCTAssertEqual(URLComponents.parse(url), URLComponents(string: string)!)
+    }
```
Comment (vanwagonet) [inline Core/CoreTests/Extensions/URLComponentsExtensionsTests.swift]: First proved that it wasn't actually `.parse()` causing the incorrect decoding.

**Comment by inst-danger** [comment]
## Latest Installable Builds

(will redirect to this page if no builds exist yet)
student: [Link](https://rebrand.ly/er3rupg) | [QR](https://qr.rebrandly.com/v1/qrcode?shortUrl=rebrand%2Ely%2Fer3rupg)
teacher: [Link](https://rebrand.ly/18fej3g) | [QR](https://qr.rebrandly.com/v1/qrcode?shortUrl=rebrand%2Ely%2F18fej3g)
parent: [Link](https://rebrand.ly/10f75u9) | [QR](https://qr.rebrandly.com/v1/qrcode?shortUrl=rebrand%2Ely%2F10f75u9)

File: Core/CoreTests/Router/RouterTests.swift [hunk 3]
```diff
@@ -81,6 +81,20 @@ class RouterTests: XCTestCase {
...
+            RouteHandler("/match") { components, _ in
+                clean = components
+                return nil
+            },
+        ]) { _, _, _ in }
+        router.route(to: .parse("/match?q=a+b&u=a%26b"), from: MockViewController())
+        XCTAssertNotNil(clean)
+        XCTAssertEqual(clean?.queryItems?[0], URLQueryItem(name: "q", value: "a b"))
+        XCTAssertEqual(clean?.queryItems?[1], URLQueryItem(name: "u", value: "a&b"))
+    }
```
Comment (vanwagonet) [inline Core/CoreTests/Router/RouterTests.swift]: then wrote a test that fails if clean uses `.query`, & ensures the + gets decoded properly.

**Comment by instructure-cx** [comment]
### Checkmarx scan completed
[Full Scan Details](https://cxmgr.inssec.info/CxWebClient/ViewerMain.aspx?scanid=1022977&projectid=17)
#### Checkmarx Scan Summary
Severity|Count
---|---
High|0
Medium|17
Low|0
Informational|0

#### Violation Summary
Severity|Count
---|---

**Comment by inst-danger** [comment]
#### Release Note: 
Fixed an issue launching Studio videos from quizzes.

#### Affected Apps: Student

[MBL-14002](https://instructure.atlassian.net/browse/MBL-14002)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Canvas iOS** | 93.49% | 93.02% | 0.47%
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Pull request merged**